In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

import warnings
warnings.filterwarnings('ignore')

from sklearn.preprocessing import MinMaxScaler, StandardScaler
from sklearn.feature_selection import RFE, RFECV
from sklearn.decomposition import PCA

from sklearn.model_selection import train_test_split, cross_val_score, GridSearchCV

from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier
import xgboost as xgb

from sklearn.metrics import confusion_matrix, accuracy_score, precision_score, recall_score,roc_curve,roc_auc_score, classification_report

from imblearn.over_sampling import RandomOverSampler

from imblearn.under_sampling import RandomUnderSampler, SMOTE,ADASYN

from imblearn.combine import SMOTETomek

In [2]:
data = pd.read_csv("telecom_churn_data.csv")
data.head()

mobile_number  circle_id  loc_og_t2o_mou  std_og_t2o_mou  loc_ic_t2o_mou  \
0     7000842753        109             0.0             0.0             0.0   
1     7001865778        109             0.0             0.0             0.0   
2     7001625959        109             0.0             0.0             0.0   
3     7001204172        109             0.0             0.0             0.0   
4     7000142493        109             0.0             0.0             0.0   

  last_date_of_month_6 last_date_of_month_7 last_date_of_month_8  \
0            6/30/2014            7/31/2014            8/31/2014   
1            6/30/2014            7/31/2014            8/31/2014   
2            6/30/2014            7/31/2014            8/31/2014   
3            6/30/2014            7/31/2014            8/31/2014   
4            6/30/2014            7/31/2014            8/31/2014   

  last_date_of_month_9   arpu_6  ...  sachet_3g_9  fb_user_6  fb_user_7  \
0            9/30/2014  197.385  ...            0        1.0        1.0   
1            9/30/2014   34.047  ...            0        NaN        1.0   
2            9/30/2014  167.690  ...            0        NaN        NaN   
3            9/30/2014  221.338  ...            0        NaN        NaN   
4            9/30/2014  261.636  ...            0        0.0        NaN   

   fb_user_8  fb_user_9   aon  aug_vbc_3g  jul_vbc_3g  jun_vbc_3g  sep_vbc_3g  
0        1.0        NaN   968        30.4         0.0      101.20        3.58  
1        1.0        NaN  1006         0.0         0.0        0.00        0.00  
2        NaN        1.0  1103         0.0         0.0        4.17        0.00  
3        NaN        NaN  2491         0.0         0.0        0.00        0.00  
4        NaN        NaN  1526         0.0         0.0        0.00        0.00  

[5 rows x 226 columns]

In [3]:
pd.set_option('display.max_columns', None)

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99999 entries, 0 to 99998
Columns: 226 entries, mobile_number to sep_vbc_3g
dtypes: float64(179), int64(35), object(12)
memory usage: 172.4+ MB


### So there are 226 columns, with 99998 rows.
Let us check for columns with major null values.

In [5]:
null_p = data.isnull().sum() / len(data) * 100
null_p

mobile_number     0.00000
circle_id         0.00000
loc_og_t2o_mou    1.01801
std_og_t2o_mou    1.01801
loc_ic_t2o_mou    1.01801
                   ...   
aon               0.00000
aug_vbc_3g        0.00000
jul_vbc_3g        0.00000
jun_vbc_3g        0.00000
sep_vbc_3g        0.00000
Length: 226, dtype: float64

In [6]:
drop_null_p = null_p[null_p>40]

In [7]:
len(drop_null_p)

40

### a total of 40 columns have null data percentage more than 40, we can remove these from data.

In [8]:
data.drop(drop_null_p.index, axis = 1, inplace = True)

In [9]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99999 entries, 0 to 99998
Columns: 186 entries, mobile_number to sep_vbc_3g
dtypes: float64(143), int64(35), object(8)
memory usage: 141.9+ MB


In [10]:
data.describe()

mobile_number  circle_id  loc_og_t2o_mou  std_og_t2o_mou  \
count   9.999900e+04    99999.0         98981.0         98981.0   
mean    7.001207e+09      109.0             0.0             0.0   
std     6.956694e+05        0.0             0.0             0.0   
min     7.000000e+09      109.0             0.0             0.0   
25%     7.000606e+09      109.0             0.0             0.0   
50%     7.001205e+09      109.0             0.0             0.0   
75%     7.001812e+09      109.0             0.0             0.0   
max     7.002411e+09      109.0             0.0             0.0   

       loc_ic_t2o_mou        arpu_6        arpu_7        arpu_8        arpu_9  \
count         98981.0  99999.000000  99999.000000  99999.000000  99999.000000   
mean              0.0    282.987358    278.536648    279.154731    261.645069   
std               0.0    328.439770    338.156291    344.474791    341.998630   
min               0.0  -2258.709000  -2014.045000   -945.808000  -1899.505000   
25%               0.0     93.411500     86.980500     84.126000     62.685000   
50%               0.0    197.704000    191.640000    192.080000    176.849000   
75%               0.0    371.060000    365.344500    369.370500    353.466500   
max               0.0  27731.088000  35145.834000  33543.624000  38805.617000   

        onnet_mou_6   onnet_mou_7   onnet_mou_8   onnet_mou_9  offnet_mou_6  \
count  96062.000000  96140.000000  94621.000000  92254.000000  96062.000000   
mean     132.395875    133.670805    133.018098    130.302327    197.935577   
std      297.207406    308.794148    308.951589    308.477668    316.851613   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        7.380000      6.660000      6.460000      5.330000     34.730000   
50%       34.310000     32.330000     32.360000     29.840000     96.310000   
75%      118.740000    115.595000    115.860000    112.130000    231.860000   
max     7376.710000   8157.780000  10752.560000  10427.460000   8362.360000   

       offnet_mou_7  offnet_mou_8  offnet_mou_9  roam_ic_mou_6  roam_ic_mou_7  \
count  96140.000000  94621.000000  92254.000000   96062.000000   96140.000000   
mean     197.045133    196.574803    190.337222       9.950013       7.149898   
std      325.862803    327.170662    319.396092      72.825411      73.447948   
min        0.000000      0.000000      0.000000       0.000000       0.000000   
25%       32.190000     31.630000     27.130000       0.000000       0.000000   
50%       91.735000     92.140000     87.290000       0.000000       0.000000   
75%      226.815000    228.260000    220.505000       0.000000       0.000000   
max     9667.130000  14007.340000  10310.760000   13724.380000   15371.040000   

       roam_ic_mou_8  roam_ic_mou_9  roam_og_mou_6  roam_og_mou_7  \
count   94621.000000   92254.000000   96062.000000   96140.000000   
mean        7.292981       6.343841      13.911337       9.818732   
std        68.402466      57.137537      71.443196      58.455762   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       0.000000   
50%         0.000000       0.000000       0.000000       0.000000   
75%         0.000000       0.000000       0.000000       0.000000   
max     13095.360000    8464.030000    3775.110000    2812.040000   

       roam_og_mou_8  roam_og_mou_9  loc_og_t2t_mou_6  loc_og_t2t_mou_7  \
count   94621.000000   92254.000000      96062.000000      96140.000000   
mean        9.971890       8.555519         47.100763         46.473010   
std        64.713221      58.438186        150.856393        155.318705   
min         0.000000       0.000000          0.000000          0.000000   
25%         0.000000       0.000000          1.660000          1.630000   
50%         0.000000       0.000000         11.910000         11.610000   
75%         0.000000       0.000000         40.960000         39.910000   
max      

In [11]:
data[['circle_id', 'loc_og_t2o_mou', 'std_og_t2o_mou', 'loc_ic_t2o_mou']].value_counts()

circle_id  loc_og_t2o_mou  std_og_t2o_mou  loc_ic_t2o_mou
109        0.0             0.0             0.0               98981
Name: count, dtype: int64

### these 4 columns also do not have any data to contribute in analysis, we can remove them

In [12]:
data.drop(['circle_id', 'loc_og_t2o_mou', 'std_og_t2o_mou', 'loc_ic_t2o_mou'], axis = 1, inplace = True)
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99999 entries, 0 to 99998
Columns: 182 entries, mobile_number to sep_vbc_3g
dtypes: float64(140), int64(34), object(8)
memory usage: 138.9+ MB


In [13]:
data.head()

mobile_number last_date_of_month_6 last_date_of_month_7  \
0     7000842753            6/30/2014            7/31/2014   
1     7001865778            6/30/2014            7/31/2014   
2     7001625959            6/30/2014            7/31/2014   
3     7001204172            6/30/2014            7/31/2014   
4     7000142493            6/30/2014            7/31/2014   

  last_date_of_month_8 last_date_of_month_9   arpu_6   arpu_7   arpu_8  \
0            8/31/2014            9/30/2014  197.385  214.816  213.803   
1            8/31/2014            9/30/2014   34.047  355.074  268.321   
2            8/31/2014            9/30/2014  167.690  189.058  210.226   
3            8/31/2014            9/30/2014  221.338  251.102  508.054   
4            8/31/2014            9/30/2014  261.636  309.876  238.174   

    arpu_9  onnet_mou_6  onnet_mou_7  onnet_mou_8  onnet_mou_9  offnet_mou_6  \
0   21.100          NaN          NaN         0.00          NaN           NaN   
1   86.285        24.11        78.68         7.68        18.34         15.74   
2  290.714        11.54        55.24        37.26        74.81        143.33   
3  389.500        99.91        54.39       310.98       241.71        123.31   
4  163.426        50.31       149.44        83.89        58.78         76.96   

   offnet_mou_7  offnet_mou_8  offnet_mou_9  roam_ic_mou_6  roam_ic_mou_7  \
0           NaN          0.00           NaN            NaN            NaN   
1         99.84        304.76         53.76            0.0           0.00   
2        220.59        208.36        118.91            0.0           0.00   
3        109.01         71.68        113.54            0.0          54.86   
4         91.88        124.26         45.81            0.0           0.00   

   roam_ic_mou_8  roam_ic_mou_9  roam_og_mou_6  roam_og_mou_7  roam_og_mou_8  \
0           0.00            NaN            NaN            NaN           0.00   
1           0.00           0.00            0.0           0.00           0.00   
2           0.00          38.49            0.0           0.00           0.00   
3          44.38           0.00            0.0          28.09          39.04   
4           0.00           0.00            0.0           0.00           0.00   

   roam_og_mou_9  loc_og_t2t_mou_6  loc_og_t2t_mou_7  loc_og_t2t_mou_8  \
0            NaN               NaN               NaN              0.00   
1           0.00             23.88             74.56              7.68   
2          70.94              7.19             28.74             13.58   
3           0.00             73.68             34.81             10.61   
4           0.00             50.31            149.44             83.89   

   loc_og_t2t_mou_9  loc_og_t2m_mou_6  loc_og_t2m_mou_7  loc_og_t2m_mou_8  \
0               NaN               NaN               NaN              0.00   
1             18.34             11.51             75.94            291.86   
2             14.39             29.34             16.86             38.46   
3             15.49            107.43             83.21             22.46   
4             58.78             67.64             91.88            124.26   

   loc_og_t2m_mou_9  loc_og_t2f_mou_6  loc_og_t2f_mou_7  loc_og_t2f_mou_8  \
0               NaN               NaN               NaN              0.00   
1             53.76              0.00              0.00              0.00   
2             28.16             24.11             21.79             15.61   
3             65.46              1.91              0.65              4.91   
4             37.89              0.00              0.00              0.00   

   loc_og_t2f_mou_9  loc_og_t2c_mou_6  loc_og_t2c_mou_7  loc_og_t2c_mou_8  \
0               NaN               NaN               NaN              0.00   
1              0.00               0.0              2.91              0.00   
2             22.24               0.0            135.54             45.76   
3              2.06               0.0              0.00              0.00   
4   

In [14]:
data[['last_date_of_month_6' , 'last_date_of_month_7', 'last_date_of_month_8', 'last_date_of_month_9']].value_counts()

last_date_of_month_6  last_date_of_month_7  last_date_of_month_8  last_date_of_month_9
6/30/2014             7/31/2014             8/31/2014             9/30/2014               98340
Name: count, dtype: int64

### these 4 columns are also not adding any value to data, its good to remove these

In [15]:
data.drop(['last_date_of_month_6' , 'last_date_of_month_7', 'last_date_of_month_8', 'last_date_of_month_9'], axis = 1, inplace = True)

In [16]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99999 entries, 0 to 99998
Columns: 178 entries, mobile_number to sep_vbc_3g
dtypes: float64(140), int64(34), object(4)
memory usage: 135.8+ MB


## There are a lot of columns with 6,7,8,9 as suffix. Lets see the set of columns for _6 suffix

In [17]:
col_6 = [col for col in data.columns if col.endswith('_6')]
col_6

['arpu_6',
 'onnet_mou_6',
 'offnet_mou_6',
 'roam_ic_mou_6',
 'roam_og_mou_6',
 'loc_og_t2t_mou_6',
 'loc_og_t2m_mou_6',
 'loc_og_t2f_mou_6',
 'loc_og_t2c_mou_6',
 'loc_og_mou_6',
 'std_og_t2t_mou_6',
 'std_og_t2m_mou_6',
 'std_og_t2f_mou_6',
 'std_og_t2c_mou_6',
 'std_og_mou_6',
 'isd_og_mou_6',
 'spl_og_mou_6',
 'og_others_6',
 'total_og_mou_6',
 'loc_ic_t2t_mou_6',
 'loc_ic_t2m_mou_6',
 'loc_ic_t2f_mou_6',
 'loc_ic_mou_6',
 'std_ic_t2t_mou_6',
 'std_ic_t2m_mou_6',
 'std_ic_t2f_mou_6',
 'std_ic_t2o_mou_6',
 'std_ic_mou_6',
 'total_ic_mou_6',
 'spl_ic_mou_6',
 'isd_ic_mou_6',
 'ic_others_6',
 'total_rech_num_6',
 'total_rech_amt_6',
 'max_rech_amt_6',
 'date_of_last_rech_6',
 'last_day_rch_amt_6',
 'vol_2g_mb_6',
 'vol_3g_mb_6',
 'monthly_2g_6',
 'sachet_2g_6',
 'monthly_3g_6',
 'sachet_3g_6']

## Looking at only og subset of columns from the entire list

In [18]:
col_og_6 = [col for col in col_6 if 'og_' in col]
col_og_6

['roam_og_mou_6',
 'loc_og_t2t_mou_6',
 'loc_og_t2m_mou_6',
 'loc_og_t2f_mou_6',
 'loc_og_t2c_mou_6',
 'loc_og_mou_6',
 'std_og_t2t_mou_6',
 'std_og_t2m_mou_6',
 'std_og_t2f_mou_6',
 'std_og_t2c_mou_6',
 'std_og_mou_6',
 'isd_og_mou_6',
 'spl_og_mou_6',
 'og_others_6',
 'total_og_mou_6']

In [19]:
data[col_og_6].head()

roam_og_mou_6  loc_og_t2t_mou_6  loc_og_t2m_mou_6  loc_og_t2f_mou_6  \
0            NaN               NaN               NaN               NaN   
1            0.0             23.88             11.51              0.00   
2            0.0              7.19             29.34             24.11   
3            0.0             73.68            107.43              1.91   
4            0.0             50.31             67.64              0.00   

   loc_og_t2c_mou_6  loc_og_mou_6  std_og_t2t_mou_6  std_og_t2m_mou_6  \
0               NaN           NaN               NaN               NaN   
1               0.0         35.39              0.23              0.00   
2               0.0         60.66              4.34             41.81   
3               0.0        183.03             26.23              2.99   
4               0.0        117.96              0.00              9.31   

   std_og_t2f_mou_6  std_og_t2c_mou_6  std_og_mou_6  isd_og_mou_6  \
0               NaN               NaN           NaN           NaN   
1              0.00               0.0          0.23           0.0   
2              1.48               0.0         47.64           0.0   
3              0.00               0.0         29.23           0.0   
4              0.00               0.0          9.31           0.0   

   spl_og_mou_6  og_others_6  total_og_mou_6  
0           NaN          NaN            0.00  
1          4.68         0.00           40.31  
2         46.56         0.45          155.33  
3         10.96         0.00          223.23  
4          0.00         0.00          127.28

In [20]:
len(col_og_6)

15

In [21]:
col_og_6.pop()

'total_og_mou_6'

In [22]:
len(col_og_6)

14

# dropping all the og columns for the 6th month as there total column is already present

In [23]:
data.drop(col_og_6, axis=1, inplace = True)

## doing the same for ic columns of 6th month

In [24]:
col_ic_6 = [col for col in col_6 if 'ic_' in col]
col_ic_6

['roam_ic_mou_6',
 'loc_ic_t2t_mou_6',
 'loc_ic_t2m_mou_6',
 'loc_ic_t2f_mou_6',
 'loc_ic_mou_6',
 'std_ic_t2t_mou_6',
 'std_ic_t2m_mou_6',
 'std_ic_t2f_mou_6',
 'std_ic_t2o_mou_6',
 'std_ic_mou_6',
 'total_ic_mou_6',
 'spl_ic_mou_6',
 'isd_ic_mou_6',
 'ic_others_6']

In [25]:
data[col_ic_6].head()

roam_ic_mou_6  loc_ic_t2t_mou_6  loc_ic_t2m_mou_6  loc_ic_t2f_mou_6  \
0            NaN               NaN               NaN               NaN   
1            0.0              1.61             17.48              0.00   
2            0.0            115.69             14.38             99.48   
3            0.0             62.08            113.96             57.43   
4            0.0            105.68            106.84              1.50   

   loc_ic_mou_6  std_ic_t2t_mou_6  std_ic_t2m_mou_6  std_ic_t2f_mou_6  \
0           NaN               NaN               NaN               NaN   
1         19.09              0.00              5.90              0.00   
2        229.56             72.41             45.18             21.73   
3        233.48             43.48              1.33              1.18   
4        214.03              0.00              1.93              0.00   

   std_ic_t2o_mou_6  std_ic_mou_6  total_ic_mou_6  spl_ic_mou_6  isd_ic_mou_6  \
0               NaN           NaN            0.00           NaN           NaN   
1               0.0          5.90           26.83          0.00          1.83   
2               0.0        139.33          370.04          0.21          0.00   
3               0.0         45.99          280.08          0.59          0.00   
4               0.0          1.93          216.44          0.00          0.00   

   ic_others_6  
0          NaN  
1         0.00  
2         0.93  
3         0.00  
4         0.48

In [26]:
len(col_ic_6)

14

In [27]:
col_ic_6.remove('total_ic_mou_6')

In [28]:
len(col_ic_6)


13

In [29]:
data.drop(col_ic_6, axis = 1, inplace=True)

In [30]:
data.shape

(99999, 151)

## Dropping all the ic and og columns apart from total for all the other 7,8 and 9 months

In [31]:
col_og = [col for col in data.columns if 'og_' in col and 'total' not in col]
col_og

['roam_og_mou_7',
 'roam_og_mou_8',
 'roam_og_mou_9',
 'loc_og_t2t_mou_7',
 'loc_og_t2t_mou_8',
 'loc_og_t2t_mou_9',
 'loc_og_t2m_mou_7',
 'loc_og_t2m_mou_8',
 'loc_og_t2m_mou_9',
 'loc_og_t2f_mou_7',
 'loc_og_t2f_mou_8',
 'loc_og_t2f_mou_9',
 'loc_og_t2c_mou_7',
 'loc_og_t2c_mou_8',
 'loc_og_t2c_mou_9',
 'loc_og_mou_7',
 'loc_og_mou_8',
 'loc_og_mou_9',
 'std_og_t2t_mou_7',
 'std_og_t2t_mou_8',
 'std_og_t2t_mou_9',
 'std_og_t2m_mou_7',
 'std_og_t2m_mou_8',
 'std_og_t2m_mou_9',
 'std_og_t2f_mou_7',
 'std_og_t2f_mou_8',
 'std_og_t2f_mou_9',
 'std_og_t2c_mou_7',
 'std_og_t2c_mou_8',
 'std_og_t2c_mou_9',
 'std_og_mou_7',
 'std_og_mou_8',
 'std_og_mou_9',
 'isd_og_mou_7',
 'isd_og_mou_8',
 'isd_og_mou_9',
 'spl_og_mou_7',
 'spl_og_mou_8',
 'spl_og_mou_9',
 'og_others_7',
 'og_others_8',
 'og_others_9']

In [32]:
data.drop(col_og, axis = 1, inplace = True)

In [33]:
data.shape

(99999, 109)

In [34]:
col_ic = [col for col in data.columns if 'ic_' in col and 'total' not in col]
col_ic

['roam_ic_mou_7',
 'roam_ic_mou_8',
 'roam_ic_mou_9',
 'loc_ic_t2t_mou_7',
 'loc_ic_t2t_mou_8',
 'loc_ic_t2t_mou_9',
 'loc_ic_t2m_mou_7',
 'loc_ic_t2m_mou_8',
 'loc_ic_t2m_mou_9',
 'loc_ic_t2f_mou_7',
 'loc_ic_t2f_mou_8',
 'loc_ic_t2f_mou_9',
 'loc_ic_mou_7',
 'loc_ic_mou_8',
 'loc_ic_mou_9',
 'std_ic_t2t_mou_7',
 'std_ic_t2t_mou_8',
 'std_ic_t2t_mou_9',
 'std_ic_t2m_mou_7',
 'std_ic_t2m_mou_8',
 'std_ic_t2m_mou_9',
 'std_ic_t2f_mou_7',
 'std_ic_t2f_mou_8',
 'std_ic_t2f_mou_9',
 'std_ic_t2o_mou_7',
 'std_ic_t2o_mou_8',
 'std_ic_t2o_mou_9',
 'std_ic_mou_7',
 'std_ic_mou_8',
 'std_ic_mou_9',
 'spl_ic_mou_7',
 'spl_ic_mou_8',
 'spl_ic_mou_9',
 'isd_ic_mou_7',
 'isd_ic_mou_8',
 'isd_ic_mou_9',
 'ic_others_7',
 'ic_others_8',
 'ic_others_9']

In [35]:
data.drop(col_ic, axis =1 , inplace = True)

In [36]:
data.shape

(99999, 70)

# Left with 70 columns now after manual feature elimination

In [37]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99999 entries, 0 to 99998
Data columns (total 70 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   mobile_number        99999 non-null  int64  
 1   arpu_6               99999 non-null  float64
 2   arpu_7               99999 non-null  float64
 3   arpu_8               99999 non-null  float64
 4   arpu_9               99999 non-null  float64
 5   onnet_mou_6          96062 non-null  float64
 6   onnet_mou_7          96140 non-null  float64
 7   onnet_mou_8          94621 non-null  float64
 8   onnet_mou_9          92254 non-null  float64
 9   offnet_mou_6         96062 non-null  float64
 10  offnet_mou_7         96140 non-null  float64
 11  offnet_mou_8         94621 non-null  float64
 12  offnet_mou_9         92254 non-null  float64
 13  total_og_mou_6       99999 non-null  float64
 14  total_og_mou_7       99999 non-null  float64
 15  total_og_mou_8       99999 non-null 

### Finding columns with missing data

In [43]:
null_p = data.isnull().sum()/len(data) * 100
null_p = null_p[null_p>1]
null_p

onnet_mou_6            3.937039
onnet_mou_7            3.859039
onnet_mou_8            5.378054
onnet_mou_9            7.745077
offnet_mou_6           3.937039
offnet_mou_7           3.859039
offnet_mou_8           5.378054
offnet_mou_9           7.745077
date_of_last_rech_6    1.607016
date_of_last_rech_7    1.767018
date_of_last_rech_8    3.622036
date_of_last_rech_9    4.760048
dtype: float64

In [38]:
data.drop_duplicates()

mobile_number   arpu_6   arpu_7   arpu_8   arpu_9  onnet_mou_6  \
0         7000842753  197.385  214.816  213.803   21.100          NaN   
1         7001865778   34.047  355.074  268.321   86.285        24.11   
2         7001625959  167.690  189.058  210.226  290.714        11.54   
3         7001204172  221.338  251.102  508.054  389.500        99.91   
4         7000142493  261.636  309.876  238.174  163.426        50.31   
...              ...      ...      ...      ...      ...          ...   
99994     7001548952   18.471   69.161   57.530   29.950         5.40   
99995     7000607688  112.201   77.811   79.081  140.835        29.26   
99996     7000087541  229.187    0.000    0.000    0.000         1.11   
99997     7000498689  322.991  303.386  606.817  731.010         0.00   
99998     7001905007  687.065    0.000    0.000    0.000        84.34   

       onnet_mou_7  onnet_mou_8  onnet_mou_9  offnet_mou_6  offnet_mou_7  \
0              NaN         0.00          NaN           NaN           NaN   
1            78.68         7.68        18.34         15.74         99.84   
2            55.24        37.26        74.81        143.33        220.59   
3            54.39       310.98       241.71        123.31        109.01   
4           149.44        83.89        58.78         76.96         91.88   
...            ...          ...          ...           ...           ...   
99994         3.36         5.91         0.00         15.19         54.46   
99995        18.13        16.06        49.49        100.83         69.01   
99996          NaN          NaN          NaN         21.04           NaN   
99997         0.00         0.00         0.00          0.00          0.00   
99998          NaN          NaN          NaN        166.46           NaN   

       offnet_mou_8  offnet_mou_9  total_og_mou_6  total_og_mou_7  \
0              0.00           NaN            0.00            0.00   
1            304.76         53.76           40.31          178.53   
2            208.36        118.91          155.33          412.94   
3             71.68        113.54          223.23          135.31   
4            124.26         45.81          127.28          241.33   
...             ...           ...             ...             ...   
99994         52.79         27.14           20.59           60.84   
99995         66.39         56.13          130.09           87.14   
99996           NaN           NaN           25.14            0.00   
99997          0.00          0.00            0.00            0.00   
99998           NaN           NaN          269.93            0.00   

       total_og_mou_8  total_og_mou_9  total_ic_mou_6  total_ic_mou_7  \
0                0.00            0.00            0.00            0.00   
1              312.44           72.11           26.83          104.23   
2              285.46          124.94          370.04          519.53   
3              352.21          362.54          280.08          216.61   
4              208.16          104.59          216.44          198.29   
...               ...             ...             ...             ...   
99994           58.71           27.14           93.09           72.88   
99995           82.46          106.41           56.11           81.38   
99996            0.00            0.00          473.69            0.00   
99997            0.00            0.00            0.81           11.09   
99998            0.00            0.00          788.73            0.00   

       total_ic_mou_8  total_ic_mou_9  total_rech_num_6  total_rech_num_7  \
0                5.44            0.00                 4                 3   
1              423.28          188.04                 4                 9   
2              395.03          517.74                 5                 4   
3               53.13          305.38                10                11   
4              338.81          205.31                 5                 6   
...               ...             ...            

## Analyzing the columns for 6th month 

In [40]:
list_col_suffix_6 = [col for col in data.columns if '_6' in col]
data[list_col_suffix_6].head()

arpu_6  onnet_mou_6  offnet_mou_6  total_og_mou_6  total_ic_mou_6  \
0  197.385          NaN           NaN            0.00            0.00   
1   34.047        24.11         15.74           40.31           26.83   
2  167.690        11.54        143.33          155.33          370.04   
3  221.338        99.91        123.31          223.23          280.08   
4  261.636        50.31         76.96          127.28          216.44   

   total_rech_num_6  total_rech_amt_6  max_rech_amt_6 date_of_last_rech_6  \
0                 4               362             252           6/21/2014   
1                 4                74              44           6/29/2014   
2                 5               168              86           6/17/2014   
3                10               230              60           6/28/2014   
4                 5               196              56           6/26/2014   

   last_day_rch_amt_6  vol_2g_mb_6  vol_3g_mb_6  monthly_2g_6  sachet_2g_6  \
0                 252        30.13        83.57             0            0   
1                  44         0.00         0.00             0            0   
2                   0         0.00         0.00             0            0   
3                  30         0.00         0.00             0            0   
4                  50         0.00         0.00             0            1   

   monthly_3g_6  sachet_3g_6  
0             1            0  
1             0            0  
2             0            0  
3             0            0  
4             0            0

### adding a new column as a mean of total recharges done in 6th and 7th month

In [55]:
data['total_rchg_6_7'] = (data['total_rech_amt_6'] + data['total_rech_amt_7'])/2
data['total_rchg_6_7']

0        307.0
1        229.0
2        241.5
3        270.0
4        273.0
         ...  
99994     85.0
99995    110.0
99996     59.5
99997    369.5
99998    355.0
Name: total_rchg_6_7, Length: 99999, dtype: float64

### Calculating the 70th percentile of this newly added column to define X

In [56]:
percentile_70 = np.percentile(data['total_rchg_6_7'],70)
percentile_70

368.5

## finding high_value_customers

In [62]:
data['high_value_customer'] = data['total_rchg_6_7'][data['total_rchg_6_7'] >= percentile_70]
data['high_value_customer'].unique()

array([   nan, 1185. ,  519. , ..., 3925. , 2915.5, 1878. ])

In [64]:
data

mobile_number   arpu_6   arpu_7   arpu_8   arpu_9  onnet_mou_6  \
0         7000842753  197.385  214.816  213.803   21.100          NaN   
1         7001865778   34.047  355.074  268.321   86.285        24.11   
2         7001625959  167.690  189.058  210.226  290.714        11.54   
3         7001204172  221.338  251.102  508.054  389.500        99.91   
4         7000142493  261.636  309.876  238.174  163.426        50.31   
...              ...      ...      ...      ...      ...          ...   
99994     7001548952   18.471   69.161   57.530   29.950         5.40   
99995     7000607688  112.201   77.811   79.081  140.835        29.26   
99996     7000087541  229.187    0.000    0.000    0.000         1.11   
99997     7000498689  322.991  303.386  606.817  731.010         0.00   
99998     7001905007  687.065    0.000    0.000    0.000        84.34   

       onnet_mou_7  onnet_mou_8  onnet_mou_9  offnet_mou_6  offnet_mou_7  \
0              NaN         0.00          NaN           NaN           NaN   
1            78.68         7.68        18.34         15.74         99.84   
2            55.24        37.26        74.81        143.33        220.59   
3            54.39       310.98       241.71        123.31        109.01   
4           149.44        83.89        58.78         76.96         91.88   
...            ...          ...          ...           ...           ...   
99994         3.36         5.91         0.00         15.19         54.46   
99995        18.13        16.06        49.49        100.83         69.01   
99996          NaN          NaN          NaN         21.04           NaN   
99997         0.00         0.00         0.00          0.00          0.00   
99998          NaN          NaN          NaN        166.46           NaN   

       offnet_mou_8  offnet_mou_9  total_og_mou_6  total_og_mou_7  \
0              0.00           NaN            0.00            0.00   
1            304.76         53.76           40.31          178.53   
2            208.36        118.91          155.33          412.94   
3             71.68        113.54          223.23          135.31   
4            124.26         45.81          127.28          241.33   
...             ...           ...             ...             ...   
99994         52.79         27.14           20.59           60.84   
99995         66.39         56.13          130.09           87.14   
99996           NaN           NaN           25.14            0.00   
99997          0.00          0.00            0.00            0.00   
99998           NaN           NaN          269.93            0.00   

       total_og_mou_8  total_og_mou_9  total_ic_mou_6  total_ic_mou_7  \
0                0.00            0.00            0.00            0.00   
1              312.44           72.11           26.83          104.23   
2              285.46          124.94          370.04          519.53   
3              352.21          362.54          280.08          216.61   
4              208.16          104.59          216.44          198.29   
...               ...             ...             ...             ...   
99994           58.71           27.14           93.09           72.88   
99995           82.46          106.41           56.11           81.38   
99996            0.00            0.00          473.69            0.00   
99997            0.00            0.00            0.81           11.09   
99998            0.00            0.00          788.73            0.00   

       total_ic_mou_8  total_ic_mou_9  total_rech_num_6  total_rech_num_7  \
0                5.44            0.00                 4                 3   
1              423.28          188.04                 4                 9   
2              395.03          517.74                 5                 4   
3               53.13          305.38                10                11   
4              338.81          205.31                 5                 6   
...               ...             ...            

## taking the subset of non null values from high_value_customer column

In [77]:
data_high_val_cust = data.dropna(subset = ['high_value_customer'])
data_high_val_cust.shape

(30011, 72)

### checking for null value counts in new dataset

In [85]:
null_p = data_high_val_cust.isnull().sum()/len(data_high_val_cust) *100
null_p = null_p[null_p>1]
null_p

onnet_mou_6            1.052947
onnet_mou_7            1.009630
onnet_mou_8            3.125521
onnet_mou_9            5.677918
offnet_mou_6           1.052947
offnet_mou_7           1.009630
offnet_mou_8           3.125521
offnet_mou_9           5.677918
date_of_last_rech_8    1.979274
date_of_last_rech_9    2.885609
dtype: float64

In [78]:
data_high_val_cust.drop('total_rchg_6_7', axis = 1, inplace = True)

In [79]:
data_high_val_cust.describe()

mobile_number        arpu_6        arpu_7        arpu_8        arpu_9  \
count   3.001100e+04  30011.000000  30011.000000  30011.000000  30011.000000   
mean    7.001223e+09    587.284404    589.135427    534.857433    488.656197   
std     6.846405e+05    442.722413    462.897814    492.259586    496.486834   
min     7.000000e+09  -2258.709000  -2014.045000   -945.808000  -1899.505000   
25%     7.000641e+09    364.161000    365.004500    289.609500    239.184000   
50%     7.001232e+09    495.682000    493.561000    452.091000    413.152000   
75%     7.001814e+09    703.922000    700.788000    671.150000    635.553000   
max     7.002411e+09  27731.088000  35145.834000  33543.624000  38805.617000   

        onnet_mou_6   onnet_mou_7   onnet_mou_8   onnet_mou_9  offnet_mou_6  \
count  29695.000000  29708.000000  29073.000000  28307.000000  29695.000000   
mean     299.184718    307.447285    276.234168    255.228092    422.380819   
std      462.202367    483.244786    471.505518    451.888513    471.096294   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%       42.960000     42.660000     32.210000     26.140000    141.410000   
50%      128.330000    127.950000    106.280000     94.080000    285.960000   
75%      357.120000    364.567500    309.980000    280.860000    527.315000   
max     7376.710000   8157.780000  10752.560000  10427.460000   8362.360000   

       offnet_mou_7  offnet_mou_8  offnet_mou_9  total_og_mou_6  \
count  29708.000000  29073.000000  28307.000000    30011.000000   
mean     428.248096    387.121245    357.382243      686.697541   
std      487.103204    480.279273    449.766130      660.356820   
min        0.000000      0.000000      0.000000        0.000000   
25%      139.587500    107.660000     88.970000      266.170000   
50%      285.485000    251.360000    226.490000      510.230000   
75%      536.827500    493.510000    454.935000      899.505000   
max     9667.130000  14007.340000  10310.760000    10674.030000   

       total_og_mou_7  total_og_mou_8  total_og_mou_9  total_ic_mou_6  \
count    30011.000000    30011.000000    30011.000000    30011.000000   
mean       709.124730      623.774684      562.667771      307.512073   
std        685.071178      685.983313      654.205634      361.159561   
min          0.000000        0.000000        0.000000        0.000000   
25%        275.045000      188.790000      136.375000       89.975000   
50%        525.580000      435.330000      383.090000      205.240000   
75%        931.050000      833.100000      757.170000      393.680000   
max      11365.310000    14043.060000    11140.110000     6798.640000   

       total_ic_mou_7  total_ic_mou_8  total_ic_mou_9  total_rech_num_6  \
count    30011.000000    30011.000000    30011.000000      30011.000000   
mean       314.875472      295.426531      277.505207         12.121322   
std        369.654489      360.343153      356.498986          9.543550   
min          0.000000        0.000000        0.000000          0.000000   
25%         98.820000       78.930000       63.620000          6.000000   
50%        211.190000      193.440000      176.190000          9.000000   
75%        396.820000      380.410000      360.695000         15.000000   
max       7279.080000     5990.710000     7785.730000        307.000000   

       total_rech_num_7  total_rech_num_8  total_rech_num_9  total_rech_amt_6  \
count      30011.000000      30011.000000      30011.000000      30011.000000   
mean          11.913465         10.225317          9.685815        697.365833   
std            9.605532          9.478572          9.043187        539.325984   
min            0.000000          0.000000          0.000000          0.000000   
25%            6.000000          4.000000          4.000000        432.000000   
50%            9.000000          8.000000          7.000000        584.000000   
75%           15.000000         13.000000         12.000000        83

# Defining Churn

In [82]:
def compare_for_churn(row):
    if (row['total_og_mou_9'] == 0 or row['total_ic_mou_9'] == 0) and (row['vol_2g_mb_9']==0 or row['vol_3g_mb_9']==0):
        return 1
    else:
        return 0
    
data_high_val_cust['churn'] = data_high_val_cust.apply(compare_for_churn, axis = 1)
data_high_val_cust.head()

mobile_number    arpu_6    arpu_7    arpu_8   arpu_9  onnet_mou_6  \
7      7000701601  1069.180  1349.850  3171.480  500.000        57.84   
8      7001524846   378.721   492.223   137.362  166.787       413.69   
13     7002191713   492.846   205.671   593.260  322.732       501.76   
16     7000875565   430.975   299.869   187.894  206.490        50.51   
17     7000187447   690.008    18.980    25.499  257.583      1185.91   

    onnet_mou_7  onnet_mou_8  onnet_mou_9  offnet_mou_6  offnet_mou_7  \
7         54.68        52.29          NaN        453.43        567.16   
8        351.03        35.08        33.46         94.66         80.63   
13       108.39       534.24       244.81        413.31        119.28   
16        74.01        70.61        31.34        296.29        229.74   
17         9.28         7.79       558.51         61.64          0.00   

    offnet_mou_8  offnet_mou_9  total_og_mou_6  total_og_mou_7  \
7         325.91           NaN          487.53          609.24   
8         136.48        108.71          508.36          431.66   
13        482.46        214.06          907.09          192.41   
16        162.76        224.39          346.81          286.01   
17          5.54         87.89         1249.53            0.81   

    total_og_mou_8  total_og_mou_9  total_ic_mou_6  total_ic_mou_7  \
7           350.16            0.00          626.46          558.04   
8           171.56          142.18          103.86           28.49   
13         1015.26          446.09          968.61          172.58   
16          233.38          255.74          364.86          381.56   
17            0.00          631.08           91.88            2.31   

    total_ic_mou_8  total_ic_mou_9  total_rech_num_6  total_rech_num_7  \
7           428.74            0.00                 5                 5   
8            16.54           34.91                19                21   
13         1144.53          631.86                 6                 4   
16          294.46          334.56                10                 6   
17            0.00           87.13                19                 2   

    total_rech_num_8  total_rech_num_9  total_rech_amt_6  total_rech_amt_7  \
7                  7                 3              1580               790   
8                 14                15               437               601   
13                11                 7               507               253   
16                 2                 1               570               348   
17                 4                10               816                 0   

    total_rech_amt_8  total_rech_amt_9  max_rech_amt_6  max_rech_amt_7  \
7               3638                 0            1580             790   
8                120               186              90             154   
13               717               353             110             110   
16               160               220             110             110   
17                30               335             110               0   

    max_rech_amt_8  max_rech_amt_9 date_of_last_rech_6 date_of_last_rech_7  \
7             1580               0           6/27/2014           7/25/2014   
8               30              36           6/25/2014           7/31/2014   
13             130             130           6/20/2014           7/22/2014   
16             130             220           6/30/2014           7/31/2014   
17              30             130           6/30/2014           7/30/2014   

   date_of_last_rech_8 date_of_last_rech_9  last_day_rch_amt_6  \
7            8/26/2014           9/30/2014                   0   
8            8/30/2014           9/30/2014                  50   
13           8/30/2014           9/26/2014                 110   
16           8/14/2014           9/10/2014                 100   
17           8/25/2014           9/29/2014                  30   

    last_day_rch_amt_7  last_day_rch_amt_8  last_day_rch_amt_9  vol_2g_mb_6  \
7   

In [86]:
data_high_val_cust['churn'].value_counts(normalize = True)

churn
0    0.898471
1    0.101529
Name: proportion, dtype: float64

### As we can see the counts above, it clearly shows there is a class imbalance with only 10percent data being about customers leaving and 90% data is about customers staying. We need to handle this to be able to build an unbiased model.

In [89]:
col_9 = [col for col in data_high_val_cust.columns if '_9' in col]

In [90]:
col_9

['arpu_9',
 'onnet_mou_9',
 'offnet_mou_9',
 'total_og_mou_9',
 'total_ic_mou_9',
 'total_rech_num_9',
 'total_rech_amt_9',
 'max_rech_amt_9',
 'date_of_last_rech_9',
 'last_day_rch_amt_9',
 'vol_2g_mb_9',
 'vol_3g_mb_9',
 'monthly_2g_9',
 'sachet_2g_9',
 'monthly_3g_9',
 'sachet_3g_9']

In [91]:
data_high_val_cust.drop(col_9, axis=1, inplace=True)

In [92]:
data_high_val_cust.shape

(30011, 56)